cpp/src/parquet/stream_reader.cc

```diff
@@ -0,0 +1,255 @@
+  CheckColumn(Type::BYTE_ARRAY, ConvertedType::UTF8);
+  ByteArray ba;
+  Read(&ba);
+  v = std::string(reinterpret_cast<const char*>(ba.ptr), ba.len);
+  return *this;
+}
+
+void StreamReader::ReadFixedLength(char* ptr, const int len) {
+  CheckColumn(Type::FIXED_LEN_BYTE_ARRAY, ConvertedType::NONE, len);
+  FixedLenByteArray flba;
+  Read(&flba);
+  std::memcpy(ptr, flba.ptr, len);
+}
+
+void StreamReader::Read(ByteArray* v) {
+  const auto& node = nodes_[column_index_];
+  auto* const reader =
+      static_cast<ByteArrayReader*>(column_readers_[column_index_++].get());
+  int64_t values_read;
+
+  reader->ReadBatch(1, nullptr, nullptr, v, &values_read);
+
+  if (values_read != 1) {
+    throw ParquetException("Failed to read value for column '" + node->name() + "'");
+  }
+}
+
+void StreamReader::Read(FixedLenByteArray* v) {
+  const auto& node = nodes_[column_index_];
+  auto* const reader =
+      static_cast<FixedLenByteArrayReader*>(column_readers_[column_index_++].get());
+  int64_t values_read;
+
+  reader->ReadBatch(1, nullptr, nullptr, v, &values_read);
+
+  if (values_read != 1) {
+    throw ParquetException("Failed to read value for column '" + node->name() + "'");
+  }
+}
+
+void StreamReader::EndRow() {
+  if (!file_reader_) {
+    throw ParquetException("StreamReader not initialized");
+  }
+  if (static_cast<std::size_t>(column_index_) < nodes_.size()) {
+    throw ParquetException("Cannot end row with " + std::to_string(column_index_) +
+                           " of " + std::to_string(nodes_.size()) + " columns read");
+  }
+  column_index_ = 0;
+
+  if (column_readers_[0]->HasNext()) {
+    return;
+  }
+  NextRowGroup();
+}
+
+void StreamReader::NextRowGroup() {
+  // Find next none-empty row group
+  while (row_group_index_ < file_metadata_->num_row_groups()) {
+    row_group_reader_ = file_reader_->RowGroup(row_group_index_);
+    ++row_group_index_;
+
+    column_readers_.resize(file_metadata_->num_columns());
+
+    for (int i = 0; i < file_metadata_->num_columns(); ++i) {
+      column_readers_[i] = row_group_reader_->Column(i);
+    }
+    if (column_readers_[0]->HasNext()) {
+      return;
+    }
+  }
+  // No more row groups found.
+  eof_ = true;
+  file_reader_.reset();
+  file_metadata_.reset();
+  row_group_reader_.reset();
+  column_readers_.clear();
+  nodes_.clear();
+}
+
+void StreamReader::CheckColumn(const Type::type physical_type,
+                               const ConvertedType::type converted_type,
+                               const int length) {
+  if (static_cast<std::size_t>(column_index_) >= nodes_.size()) {
+    if (eof_) {
+      throw ParquetException("EOF reached");
+    }
+    throw ParquetException("Column index out-of-bounds.  Index " +
+                           std::to_string(column_index_) + " is invalid for " +
+                           std::to_string(nodes_.size()) + " columns");
+  }
+  const auto& node = nodes_[column_index_];
```

Comment:
for commiter to way in.  Should auto be spelled out to type (see conversation on previous commits).